app/services/carto/visualization_backup_service.rb

```
@@ -0,0 +1,39 @@
+require_dependency 'carto/visualizations_export_service_2'
+
+module Carto
+  module VisualizationBackupService
+    include Carto::VisualizationsExportService2Exporter
+    include Carto::VisualizationsExportService2Importer
+
+    def create_visualization_backup(visualization:, category:, with_mapcaps: true, with_password: true)
+      export_json = export_visualization_json_hash(
+        visualization.id,
+        visualization.user,
+        with_mapcaps: with_mapcaps,
+        with_password: with_password
+      )
+      Carto::VisualizationBackup.create!(
+        user_id: visualization.user.id,
+        visualization_id: visualization.id,
+        category: category,
+        export: export_json
+      )
+    rescue StandardError => exception
+      # The backup should not break the flow
+      CartoDB::Logger.error(
+        message: 'Error creating a visualization backup',
+        exception: exception,
+        visualization: visualization
+      )
+    end
+
+    def restore_visualization_backup(visualization_backup_id)
+      visualization_backup = Carto::VisualizationBackup.find(visualization_backup_id)
+      user = Carto::User.find(visualization_backup.user_id)
```

Maybe raise an error if a visualization with the same ID as the backup already exists?

---

I have changed it and now it always throws following the advice of @gonzaloriestra https://github.com/CartoDB/cartodb/pull/14764#discussion_r268059814

---

^ I have updated the link to the comment

---

Yes, it should fail anyway, but I think @javitonino suggests to add a more explicit error message for that specific case.

---

Something like this:
```
raise 'The visualization to restore already exists' if Carto::Visualization.find(visualization_backup.visualization_id)
```